spec/app/ics-100-atomic-swap/README.md

```diff
@@ -0,0 +1,651 @@
+  abortTransactionUnless(counterpartyVersion === "ics100-1")
+  
+  // allocate an escrow address
+  channelEscrowAddresses[channelIdentifier] = newAddress()
+  
+  return "ics100-1", nil
+}
+```
+
+```typescript
+function onChanOpenAck(
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  counterpartyChannelIdentifier: Identifier,
+  counterpartyVersion: string
+) {
+  // port has already been validated
+  // assert that counterparty selected version is "ics100-1"
+  abortTransactionUnless(counterpartyVersion === "ics100-1");
+}
+```
+
+```typescript
+function onChanOpenConfirm(portIdentifier: Identifier, channelIdentifier: Identifier) {
+  // accept channel confirmations, port has already been validated, version has already been validated
+}
+```
+
+```typescript
+function onChanCloseInit(portIdentifier: Identifier, channelIdentifier: Identifier) {
+  // always abort transaction
+  abortTransactionUnless(FALSE);
+}
+```
+
+```typescript
+function onChanCloseConfirm(portIdentifier: Identifier, channelIdentifier: Identifier) {
+  // no action necessary
+}
+```
+
+#### Packet relay
+
+`sendAtomicSwapPacket` must be called by a transaction handler in the module which performs appropriate signature checks, specific to the account owner on the host state machine.
+
+```typescript
+function sendAtomicSwapPacket(
+  swapPacket: AtomicSwapPacketData, 
+  sourcePort: string,
+  sourceChannel: string,
+  timeoutHeight: Height,
+  timeoutTimestamp: uint64
+) {
+  // send packet using the interface defined in ICS4
+  handler.sendPacket(
+    getCapability("port"),
+    sourcePort,
+    sourceChannel,
+    timeoutHeight,
+    timeoutTimestamp,
+    swapPacket.getBytes() // Should be proto marshalled bytes.
+  )
+}
+```
+
+`onRecvPacket` is called by the routing module when a packet addressed to this module has been received. 
+
+```typescript
+function onRecvPacket(packet channeltypes.Packet) {
+  const swapPacket: AtomicSwapPacketData = packet.data
+  
+  AtomicSwapPacketAcknowledgement ack = AtomicSwapPacketAcknowledgement{true, null}
+  
+  switch swapPacket.type {
+    case TYPE_MSG_MAKE_SWAP:
+      const makeMsg = protobuf.decode(swapPaket.data)
+        
+      // check if buyToken is a valid token on the taker chain, could be either native or ibc token
+      const supply = bank.getSupply(makeMsg.buyToken.denom)
+      abortTransactionUnless(supply > 0)
+        
+      // create and save order on the taker chain.
+      const order = OrderBook.createOrder(makeMsg)
+      order.status = Status.SYNC
+      order.portIdOnTakerChain = packet.destinationPort
+      order.channelIdOnTakerChain = packet.destinationChannel
+      // saves order to store
+      const err = privateStore.set(orderPath(packet.destinationChannel, order.orderId), order)
+      if (err != null) {
+        ack = AtomicSwapPacketAcknowledgement{false, "dailed to save the order on taker chain"}
+      }
+      break;
+    case TYPE_MSG_TAKE_SWAP:
```

Comment:
I'm considering removing "Status.SYNC" to make it more flexible to sequence of packet delivery, What's your opinion?

Reply:
Another option is to remove the validation so that orders can still be processed.

Would require adjustment on the ACK of the make order so that it doesn’t overwrite Status in case of ordering.

Reply:
If you do it that way, you will have to revert this logic if the MAKE order comes back unsuccessfully. I actually think its just simpler to leave it as is. It should just be documented

Reply:
OK，Let's leave it as it is